### PR TITLE
Fixing memory leak in distributed pipeline

### PR DESCRIPTION
**Pull request by mrzzd**

### CODE REVIEW

File: fairscale/experimental/nn/distributed_pipeline/partition_handler.py [hunk 1]
```diff
@@ -275,14 +275,19 @@ def run_pipeline(self, pipeline_record_rref: rpc.RRef) -> Optional[Tensor]:
...
-            result = microbatch.gather(pipeline_record.batches)
-            assert len(result) == 1
-            result = result[0]
+            gather_result = microbatch.gather(pipeline_record.batches)
+            assert len(gather_result) == 1
+            result = gather_result[0]
             s0 = current_stream(result.device)
             if is_cuda(s0):
                 # TODO. Investigate why this is needed and remove it if possible.
                 as_cuda(s0).synchronize()
```
Comment (SciPioneer) [inline fairscale/experimental/nn/distributed_pipeline/partition_handler.py:287]: curious, what's the difference between this `synchronize` and `torch.cuda.syncrhonize(result.device)`?

Comment (mrzzd) [reply fairscale/experimental/nn/distributed_pipeline/partition_handler.py:287]: this one only synchronizes with specified stream. torch.cuda.syncrhonize waits for all operators in all streams of device.

File: fairscale/experimental/nn/distributed_pipeline/partition_handler.py [hunk 2]
```diff
@@ -275,14 +275,19 @@ def run_pipeline(self, pipeline_record_rref: rpc.RRef) -> Optional[Tensor]:
...
-            return result
 
-        return None
+        # TODO: There seems to be a memory leak that is solved by following line.
```
Comment (SciPioneer) [inline fairscale/experimental/nn/distributed_pipeline/partition_handler.py:289]: Did you confirm the memory footprint reduction by profiling?

File: fairscale/experimental/nn/distributed_pipeline/partition_handler.py [hunk 3]
```diff
@@ -275,14 +275,19 @@ def run_pipeline(self, pipeline_record_rref: rpc.RRef) -> Optional[Tensor]:
         pipeline_record = pipeline_record_rref.local_value()
         self.run(pipeline_record)
 
+        result: Optional[Tensor] = None
+
         if not pipeline_record.consumers:
-            result = microbatch.gather(pipeline_record.batches)
-            assert len(result) == 1
-            result = result[0]
+            gather_result = microbatch.gather(pipeline_record.batches)
```
Comment (SciPioneer) [inline fairscale/experimental/nn/distributed_pipeline/partition_handler.py:281]: nit: `gather_results` sounds a bit better.

**Review by SciPioneer — APPROVED** [review]

File: fairscale/experimental/nn/distributed_pipeline/partition_handler.py (hunk 2 above)
Comment (mrzzd) [reply fairscale/experimental/nn/distributed_pipeline/partition_handler.py:289]: Before the change nvidia-smi was showing continues increase in memory consumption, and after the fix memory size on gpu was stabilized.

File: fairscale/experimental/nn/distributed_pipeline/partition_handler.py (hunk 3 above)
Comment (mrzzd) [reply fairscale/experimental/nn/distributed_pipeline/partition_handler.py:281]: hmmm, results might misleading in other ways. the function has only one object returning (though it is a list).

Comment (SciPioneer) [reply fairscale/experimental/nn/distributed_pipeline/partition_handler.py:281]: Well, you assigned `gather_results[0]` to `result`, so it should be clear. It's up to you.

**Pull request merged**